doc: document nixpkgs-fmt rules (#121306)
NixOS/nixpkgs: doc/contributing/coding-conventions.chapter.md

## Syntax {#sec-syntax}

The key words _must_, _must not_, _required_, _shall_, _shall not_, _should_, _should not_, _recommended_, _may_, and _optional_ in this section are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119). Only _emphasized_ words are to be interpreted in this way.

The nixpkgs repository is formatted with [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt/).

A non-exhaustive list of formatting rules follows. When in doubt, the tool is authoritative.

> - Use 2 spaces of indentation in Nix expressions, 4 spaces in shell scripts.

Reviewer: ```suggestion
- Use 2 spaces of indentation in Nix files and 4 spaces in shell scripts.
```

Author: I think the previous formulation better captures that there can be Bash expression in nix files.

> - Do not use tab characters, i.e. configure your editor to use soft tabs. Everybody has different tab settings so it’s asking for trouble.

Reviewer: ```suggestion
- Do not use tab characters, i.e. configure your editor to use soft tabs or replace tabs with two spaces. Please refer to your editors manual on exact details. Everybody has different tab settings so it’s asking for trouble.
```

Author: What do you mean here with "replace tabs with two spaces"? When should a user do that, as opposed to configuring their editor?

- Use `lowerCamelCase` for variable names, not `UpperCamelCase`. Note, this rule does not apply to package attribute names, which instead follow the rules in <xref linkend="sec-package-naming"/>.

- Empty lists and attribute sets _must_ contain exactly one space

```nix
[ ]

# and

{ }
```

> not

Reviewer: ```suggestion
instead of
```

```nix
[]

# or

{  }
```

- There _must_ be a space between list/set delimiters and their elements

``` nix
[ 1 2 3 ]

# and

{ nixpkgs ? { } }
```

> not

Reviewer: ```suggestion
instead of
```

  ```nix
  { arg1, arg2 }:
    assert system == "i686-linux";
      with arg1;
        stdenv.mkDerivation { ...
  ```

  When functions, `assert`s and `with`s are not top-level, the normal, 2-space indentation rules apply

  ```nix
  {
    x = assert true;
      false;
    y = with "foo";
      "bar";
    z = x: "baz";
  }
  ```

- For `let ... in` expressions, the bindings _must_ be indented, whereas the expression _must not_ be indented and _must_ be on its own line

  ```nix
  let
    x = 1;
    y = 2;
  in
  x + y
  ```

  not

  ```nix
  let
    x = 1;
    y = 2;
  in
    x + y

  # or

  let
    x = 1;
    y = 2;
  in x + y
  ```

  For single bindings, more compact forms are accepted

  ```nix
  let x = 1; in x

  let x = 1;
  in x

  let x = 1;
  in
  x
  ```

- The body of multi-line strings _must_ be indented relative to the bound variable

  ```nix
  x = ''
    foo
  '';
  ```

>   not

Reviewer: ```suggestion
  instead of
```

- Arguments should be listed in the order they are used, with the exception of `lib`, which always goes first.

> - The top-level `lib` must be used in the master and 21.05 branches over its alias `stdenv.lib` as it now causes evaluation errors when aliases are disabled (which is the case for the `ofborg` GitHub bot).

Reviewer: ```suggestion
- The top-level `lib` must be used in the master and 21.05 release branche over its alias `stdenv.lib` as it now causes evaluation errors when aliases are disabled (which is the case for the `ofborg` bot).
```

Author: Do you mean this to say "master and 21.05 release branches" or "master branch and 21.05 release branch"? If it's the former, I don't think master is a RB, and if it's the latter, I'm not sure it's necessary.
